src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -0,0 +1,483 @@
+#define NET8_0_OR_GREATER
+#endif
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+#nullable enable
+
+namespace System.Reflection.Metadata
+{
+    // TODO: add proper debugger display stuff
+#if SYSTEM_PRIVATE_CORELIB
+    internal
+#else
+    public
+#endif
+    ref struct TypeNameParser
+    {
+        private const string EndOfTypeNameDelimiters = "[]&*,+";
+#if NET8_0_OR_GREATER
+        private static readonly SearchValues<char> _endOfTypeNameDelimitersSearchValues = SearchValues.Create(EndOfTypeNameDelimiters);
+#endif
+        private static readonly TypeNameParserOptions _defaults = new();
+        private readonly bool _throwOnError;
+        private readonly TypeNameParserOptions _parseOptions;
+        private ReadOnlySpan<char> _inputString;
+
+        private TypeNameParser(ReadOnlySpan<char> name, bool throwOnError, TypeNameParserOptions? options) : this()
+        {
+            _inputString = name;
+            _throwOnError = throwOnError;
+            _parseOptions = options ?? _defaults;
+        }
+
+        public static TypeName? Parse(ReadOnlySpan<char> typeName, bool allowFullyQualifiedName = true, bool throwOnError = true, TypeNameParserOptions? options = default)
+        {
+            ReadOnlySpan<char> trimmedName = TrimStart(typeName); // whitespaces at beginning are always OK
+            if (trimmedName.IsEmpty)
+            {
+                // whitespace input needs to report the error index as 0
+                return ThrowInvalidTypeNameOrReturnNull(throwOnError, 0);
+            }
+
+            int recursiveDepth = 0;
+            TypeNameParser parser = new(trimmedName, throwOnError, options);
+            TypeName? parsedName = parser.ParseNextTypeName(allowFullyQualifiedName, ref recursiveDepth);
+
+            if (parsedName is not null && parser._inputString.IsEmpty) // unconsumed input == error
+            {
+                return parsedName;
+            }
+            else if (!throwOnError)
+            {
+                return null;
+            }
+
+            // there was an error and we need to throw
+#if !SYSTEM_PRIVATE_CORELIB
+            if (recursiveDepth >= parser._parseOptions.MaxRecursiveDepth)
+            {
+                throw new InvalidOperationException("SR.RecursionCheck_MaxDepthExceeded");
+            }
+#endif
+            int errorIndex = typeName.Length - parser._inputString.Length;
+            return ThrowInvalidTypeNameOrReturnNull(throwOnError, errorIndex);
+
+            static TypeName? ThrowInvalidTypeNameOrReturnNull(bool throwOnError, int errorIndex = 0)
+            {
+                if (!throwOnError)
+                {
+                    return null;
+                }
+
+#if SYSTEM_PRIVATE_CORELIB
+                throw new ArgumentException(SR.Arg_ArgumentException, $"typeName@{errorIndex}");
+#else
+                throw new ArgumentException("SR.Argument_InvalidTypeName");
```

Comment:
```suggestion
                throw new ArgumentException(SR.Argument_InvalidTypeName);
```

Perhaps you meant to reference a resource here?